File: web/src/admin/admin_api.ts
const ADMIN_API_URL = "http://localhost:8080/api/admin";

/*
 * These are the only statuses that can exist
 * after a report has already been approved.
 */
export type PublicItemStatus =
  | "Unclaimed"
  | "Pending Claim"
  | "Claimed";

/*
 * Get all reports.
 */
export async function getAllReports() {
  const response = await fetch(`${ADMIN_API_URL}/reports`);

  if (!response.ok) {
    const message = await response.text();

    throw new Error(
      message || "Failed to fetch reports."
    );
  }

  return response.json();
}

/*
 * Get all reported items.
 */
export async function getReportedItems() {
  const response = await fetch(
    `${ADMIN_API_URL}/reported-items`
  );

  if (!response.ok) {
    const message = await response.text();

    throw new Error(
      message || "Failed to fetch reported items."
    );
  }

  return response.json();
}

/*
 * ==========================================================
 * APPROVE REPORT
 *
 * Under Review
 *        ↓
 *   Unclaimed
 * ==========================================================
 */
export async function approveReport(id: string) {
  const response = await fetch(
    `${ADMIN_API_URL}/reports/${encodeURIComponent(id)}/approve`,
    {
      method: "PUT",
    }
  );

  if (!response.ok) {
    const message = await response.text();

    throw new Error(
      message || "Failed to approve report."
    );
  }

  return response.json();
}

/*
 * ==========================================================
 * REJECT REPORT
 *
 * Under Review
 *        ↓
 *   Rejected
 * ==========================================================
 */
export async function rejectReport(id: string) {
  const response = await fetch(
    `${ADMIN_API_URL}/reports/${encodeURIComponent(id)}/reject`,
    {
      method: "PUT",
    }
  );

  if (!response.ok) {
    const message = await response.text();

    throw new Error(
      message || "Failed to reject report."
    );
  }

  return response.json();
}

/*
 * ==========================================================
 * UPDATE STATUS
 *
 * Only used AFTER a report has already been
 * approved.
 *
 * Allowed values:
 *
 * Unclaimed
 * Pending Claim
 * Claimed
 * ==========================================================
 */
export async function updateReportStatus(
  id: string,
  status: PublicItemStatus
) {
  const response = await fetch(
    `${ADMIN_API_URL}/reports/${encodeURIComponent(id)}/status`,
    {
      method: "PUT",
      headers: {
        "Content-Type": "application/json",
      },
      body: JSON.stringify({
        status,
      }),
    }
  );

  if (!response.ok) {
    const message = await response.text();

    throw new Error(
      message || "Failed to update report status."
    );
  }

  return response.json();
}

File: mobile/app/src/main/java/cit/edu/garol/campus/features/admin/api/AdminApi.kt
package cit.edu.garol.campus.features.admin.api

import cit.edu.garol.campus.features.admin.model.AdminLoginRequest
import cit.edu.garol.campus.features.admin.model.AdminLoginResponse
import cit.edu.garol.campus.features.admin.model.AdminReportItem
import cit.edu.garol.campus.features.admin.model.AdminStatusRequest
import retrofit2.Response
import retrofit2.http.Body
import retrofit2.http.GET
import retrofit2.http.POST
import retrofit2.http.PUT
import retrofit2.http.Path

interface AdminApi {

    /*
     * ==========================================================
     * ADMIN LOGIN
     * ==========================================================
     */

    @POST("api/admin/login")
    suspend fun loginAdmin(
        @Body request: AdminLoginRequest
    ): Response<AdminLoginResponse>

    /*
     * ==========================================================
     * REPORTS
     * ==========================================================
     */

    @GET("api/admin/reported-items")
    suspend fun getReportedItems(): Response<List<AdminReportItem>>

    /*
     * ==========================================================
     * APPROVE REPORT
     *
     * Under Review -> Unclaimed
     * ==========================================================
     */

    @PUT("api/admin/reports/{id}/approve")
    suspend fun approveReport(
        @Path("id") reportId: String
    ): Response<AdminReportItem>

    /*
     * ==========================================================
     * REJECT REPORT
     *
     * Under Review -> Rejected
     * ==========================================================
     */

    @PUT("api/admin/reports/{id}/reject")
    suspend fun rejectReport(
        @Path("id") reportId: String
    ): Response<AdminReportItem>

    /*
     * ==========================================================
     * UPDATE PUBLIC STATUS
     *
     * Unclaimed
     * Pending Claim
     * Claimed
     * ==========================================================
     */

    @PUT("api/admin/reports/{id}/status")
    suspend fun updateReportStatus(
        @Path("id") reportId: String,
        @Body request: AdminStatusRequest
    ): Response<AdminReportItem>

    @GET("api/reports/public")
    suspend fun getPublicReports():
            Response<List<AdminReportItem>>
}